bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1295,4 +1298,127 @@ public boolean equals(Object otherPair) {
             return true;
         }
     }
+
+    /**
+     * Checks whether a given {@link Byte} has an implicit initial value or not.
+     * @param type {@link Byte} to be analyzed.
+     * @return whether there's an implicit initial value or not.
+     */
+    public static boolean hasImplicitInitialValue(BType type) {
+        return hasImplicitInitialValue(type, new ArrayList<>());
+    }
+
+    private static boolean hasImplicitInitialValue(BType type, List<BType> unAnalyzedTypes) {
+        if (type == null) {
+            return true;
+        }
+        if (type.getTag() < TypeTags.RECORD_TYPE_TAG) {
+            return true;
+        }
+        switch (type.getTag()) {
+            case TypeTags.TYPEDESC_TAG:
+            case TypeTags.STREAM_TAG:
+            case TypeTags.MAP_TAG:
+            case TypeTags.ANY_TAG:
+                return true;
+            case TypeTags.ARRAY_TAG:
+                return hasImplicitInitialValue(((BArrayType) type).getElementType());
+            case TypeTags.FINITE_TYPE_TAG:
+                return analyzeFiniteType((BFiniteType) type);
+            case TypeTags.OBJECT_TYPE_TAG:
+                return analyzeObjectType((BObjectType) type);
+            case TypeTags.RECORD_TYPE_TAG:
+                return analyzeRecordType((BRecordType) type, unAnalyzedTypes);
+            case TypeTags.TUPLE_TAG:
+                BTupleType tupleType = (BTupleType) type;
+                return tupleType.getTupleTypes().stream().allMatch(TypeChecker::hasImplicitInitialValue);
+            case TypeTags.UNION_TAG:
+                return analyzeUnionType((BUnionType) type);
+            default:
+                return false;
+        }
+    }
+
+    private static boolean analyzeUnionType(BUnionType type) {
+        // NIL is a member.
+        if (type.isNullable()) {
+            return true;
+        }
+        // All members are of same type.
+        Iterator<BType> iterator = type.getMemberTypes().iterator();
+        BType firstMember;
+        for (firstMember = iterator.next(); iterator.hasNext(); ) {
+            if (!isSameType(firstMember, iterator.next())) {
+                return false;
+            }
+        }
+        // Control reaching this point means there is only one type in the union.
+        return BTypes.isValueType(firstMember) && hasImplicitInitialValue(firstMember);
+    }
+
+    private static boolean analyzeRecordType(BStructureType type, List<BType> unanalyzedTypes) {
+        if (unanalyzedTypes.contains(type)) {
+            return true;
+        }
+        unanalyzedTypes.add(type);
+        return type.getFields().values().stream().allMatch(f -> hasImplicitInitialValue(f.type, unanalyzedTypes));
```

Comment:
According to the spec, for mapping types the filler value is `{ }` if that is a valid constructor for the type.

Does that mean we need to consider records with fields with default values, even though they have no filler values, as records with a filler value?

```ballerina
type Bar object {
    function __init(string s) {
        // ....
    }
};

type Bar record {
    Foo f = new("test");
};
```
Since `{}` is a valid constructor for `Bar`, shouldn't we allow `Bar[]`

Reply:
`Bar[]` is ok in my view. The following was allowed before as well.
```ballerina
type Foo object {
    string f;

    function __init() {
        self.f = "foo";
    }
};
```
`Foo[]` was allowed since `Foo` has a no-arg `__init()` method.

Reply:
Yeah, for objects the same rule regarding the no-arg init still applies. My point is that for records we may also need to check for default values?

Reply:
+1 for allowing Bar[]
When this check was first implemented, rec field default values were not fully supported (I think we only had default literals or something) , that's why it's not checked here.
Is record field default value support completed now?